File: ws/src/robot_firmware/robot_firmware/simple_serial_receiver.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
import serial


class SimpleSerialReceiver(Node):
    def __init__(self):
        super().__init__("simple_serial_receiver")
        self.declare_parameter("port", "/dev/ttyACM0")
        self.declare_parameter("baud_rate", 115200)       

        self.__port = self.get_parameter("port").value
        self.__baud_rate: int = self.get_parameter("baud_rate").value

        self.pub_ = self.create_publisher(String, 'serial_receiver', 10)
        self.__frecuency = 0.01
        self.__arduino = serial.Serial(port=self.__port, baudrate=self.__baud_rate)
        self.__timer = self.create_timer(self.__frecuency, self.timerCallback)


    def timerCallback(self):
        if rclpy.ok() and self.__arduino.is_open:
            data = self.__arduino.readline()
            try:
                data.decode()
            except Exception as err:
                return
            msg = String()
            msg.data = str(data)
            self.pub_.publish(msg)

def main():
    rclpy.init()
    simple_serial_receiver = SimpleSerialReceiver()
    rclpy.spin(simple_serial_receiver)
    simple_serial_receiver.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: ws/src/robot_firmware/src/simple_serial_receiver.cpp
#include <functional>
#include <libserial/SerialPort.h>
#include <libserial/SerialPortConstants.h>
#include <memory>
#include <rclcpp/executors.hpp>
#include <rclcpp/logging.hpp>
#include <rclcpp/node.hpp>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp/subscription.hpp>
#include <rclcpp/timer.hpp>
#include <rclcpp/utilities.hpp>
#include <std_msgs/msg/detail/string__struct.hpp>
#include <std_msgs/msg/string.hpp>
#include <string>

using std::placeholders::_1;
using namespace std::chrono_literals;

class SimpleSerialReceiver : public rclcpp::Node {
public:
  SimpleSerialReceiver() : Node("simple_serial_receiver") {
    declare_parameter<std::string>("port", "/dev/ttyACM0");
    std::string port_ = get_parameter("port").as_string();
    pub_ = create_publisher<std_msgs::msg::String>("serial_receiver", 10);
    timer_ = create_wall_timer(0.01s,  std::bind(&SimpleSerialReceiver::timerCallback, this));

    arduino_.Open(port_);
    arduino_.SetBaudRate(LibSerial::BaudRate::BAUD_115200);
  }

  ~SimpleSerialReceiver() { arduino_.Close(); };

  void timerCallback() {
   auto message = std_msgs::msg::String();
    if(rclcpp::ok()&&arduino_.IsDataAvailable()){
        arduino_.ReadLine(message.data);
    }
    pub_->publish(message);
  }

private:
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  LibSerial::SerialPort arduino_;
};

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimpleSerialReceiver>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}